### DISCUSSION TITLE
Mobile / API Security / Best Practices

### DISCUSSION THREAD

**Discussion by nynymike** [MobileNativeFoundation/discussions#12] [Proposals]
Hi guys. As you know, mobile native applications call a lot of API's. I end up having a lot of conversations about how to use OAuth access tokens and refresh tokens to implement various security requirements. 

One of the challenges we are dealing with is advising customers how to best architect mobile applications. [RFC 8252](https://tools.ietf.org/html/rfc8252) is a fantastic reference around OAuth security for mobile applications, especially with regard to essential mitigations with regard to the use of a third-party federated identity provider. 

However, I wonder if the productivity of mobile app development could be increased if there were more playbooks or patterns for achieving security objectives end-to-end, from API design down to implementation in the mobile application. 

As one of the leaders at the [Janssen Project](https://jans.io), I'd be interested to work on such a topic.

**Comment by shepting** [comment]
Does Janssen have any mobile libraries?

**Reply by donnie-jp** [reply]
Yeah, providing mobile SDKs would be the easiest way to get apps to use the feature correctly.

**Reply by nynymike** [reply]
Janssen is an OAuth Server. Mobile clients would use OAuth SDKs or [AppAuth-Android](https://github.com/openid/AppAuth-Android), [AppAuth-iOS](https://github.com/openid/AppAuth-iOS) or [AppAuth-JS](https://github.com/openid/AppAuth-JS)

But that is only half the issue... there has to be coordination between the mobile client, api gateway layer, and the OAuth Authorization Server.  These three components work together to implement an end-to-end solution.

**Comment by freak4pc** [comment]
Definitely a less written-about topic that's very interesting. 

I think that if we're taking about company-internals, the easiest thing to do is abstract the OAuth mechanism well enough so its not noticeable in the day-to-day. Your post does seem to suggest it's more about external clients which is a different play, for sure.

I will say that security in iOS is a bit of a no-man's land, indeed. There are many enterprise companies that purchase products that teach developers how to apply proper security to their applications, and the iOS/Android portions usually end up looking quite mediocre in this somewhat backend/frontend-central.

I'd be happy for a resource such as the one you suggested, personally.

**Comment by manijshrestha** [comment]
Per [RFC 8253](https://tools.ietf.org/html/rfc8252), authorization requests should be made through an external browser (primary browser). I have seen it to be a hard problem on Android. Primarily due to different WebKit and browser behaviors by different manufacturers. I would like to hear about the success stories of implementing a browser-based auth mechanism for a native android app.

**Reply by nynymike** [reply]
There is a proposal by Travis Spencer to implement a new approach that is non-browser based called the "Hypermedia Authentication API". You can read about it here: https://travisspencer.com/articles/hypermedia-api-resources/

I really like this idea, and Gluu is considering contributed an implementation of this API to the Janssen Auth Server. 

Net-net, I agree that reliance on browser SDKs makes it hard for mobile developers.

**Comment by Exadra37** [comment]
> As you know, mobile native applications call a lot of API's. I end up having a lot of conversations about how to use OAuth access tokens and refresh tokens to implement various security requirements.

I work as a Developer Advocate for Security in mobile apps and APIs, and I see developers relying on User Authentication mechanisms to solve the very hard problem of locking down the mobile app to it's API backend,. Like I do in all [my StackOverflow](https://stackoverflow.com/users/6454622/exadra37) security answers I would also like to clarify here the misconception that developers of any seniority may have or not regarding the difference between **who** vs **what** is doing the API request.

![man-in-themiddle-attack](https://user-images.githubusercontent.com/1745456/112131948-02b0f580-8bc2-11eb-998a-d3ffa13ac783.png)

The Intended Communication Channel represents your mobile being used as you expected, by a legit user without any malicious intentions, using an untampered version of your mobile app, and communicating directly with your API server without being man in the middle attacked.

The actual channel may represent several different scenarios, like a legit user with malicious intentions that may be using a repackaged version of your mobile app, a hacker using the genuine version of you mobile app while man in the middle attacking it to understand how the communication between the mobile app and the API server is being done in order to be able to automate attacks against your API. Many other scenarios are possible, but we will not enumerate each one here.

Usually mobile apps need to use some sort of key/secret to identify themselves to the API backend, like an API key or a user authorization token, or they will try to use unique device identifiers to allow for a friction-less user experience were they don't require user authentication while keeping track of each different user by some sort of means, but some simply don’t use anything.

No matter **what** type of approach is used to identify the mobile app to the API backend it’s important that the API backend is able to distinguish between **who** is in the request versus **what** is doing the request. 

The **who** is the user of the mobile app that we can authenticate, authorize and identify in several ways, like using OpenID Connect, OAUTH2 flows or any other user authentication mechanism.

The **what** is the thing making the request to the API server. Is it really a genuine instance of your mobile app(using the intended communication channel), or is a bot, an automated script or an attacker(like the black hat one above) manually poking around your API backend with a tool like Insomnia, Postman, cURL, etc.?

So, in a nutshell OAUTH2 is not enough to give an high degree of confidence to the API server that the current request is indeed from **what** it expects, a genuine and unmodified instance of the mobile app.
